content/docs/document-lifecycle.md

```diff
@@ -0,0 +1,55 @@
+---
+title: 'Document lifecycle'
+---
+
+A `LangiumDocument` goes through seven different states during its lifecycle:
+    * `Parsed` when an AST has been generated from the content of the document.
+    * `IndexedContent` when the AST nodes have been processed by the `IndexManager`.
+    * `Processed` when the pre-processing steps have been completed.
+    * `Linked` when the `Linker` has resolved cross-references.
+    * `IndexedReferences` when the references have been indexed by the `IndexManager`.
+    * `Validated` when the document has been validated by the `DocumentValidator`.
+    * `Changed` when the document has been modified.
+
+## Creation of LangiumDocuments
+When the workspace is initialized, all files having an extension matching those defined in `langium-config.json` will be collected. During collection, the `WorkspaceManager` relies on the `LangiumDocuments` service to check if a `LangiumDocument` exist for a given URI. If no instance of a `LangiumDocument` can be found, the `LangiumDocumentFactory` service will create a new instance of `LangiumDocument`. 
+
+Files in the workspace are inherently instances of `TextDocument` as implemented by VS Code's language server. These `TextDocument`s hold the content of the respective file as a `string`. Conversely, a `LangiumDocument` does not hold the content of document as a string but as an AST. In other words, the creation of a `LangiumDocument` by the `LangiumDocumentFactory` service is accompanied by the parsing of the content of a `TextDocument` into a given AST. During the creation of a `LangiumDocument` (i.e. after the document has been parsed), its state is set to `Parsed`.
+
+Once all `LangiumDocument`s have been created, the `DocumentBuilder` service will sequentially process each `LangiumDocument` as described below.
+
+## Indexing of the AST
+Initial indexing of the AST is executed on `LangiumDocument`s with the state `Parsed`. The default `AstNodeDescriptionProvider` service creates an `AstNodeDescription` for the root node (i.e. the node created by parsing the entry rule) and each named `AstNode` directly descending from the root node. This `AstNodeDescription` contains the `type` of the node, its identifier (i.e. the `name` property), the uri of the document where the node is located, and the location of the node inside of the document. The generated array of `AstNodeDescription`s is then mapped into a `simpleIndex`. This index makes accessible named nodes from a `LangiumDocument` to other `LangiumDocument`s in the same workspace.
```

Comment:
`simpleIndex` is an implementation detail of `IndexManager` and should not be mentioned here.